### PR TITLE
[Bind] Bind2

### PR DESCRIPTION
**Pull request by rsetaluri**

### CODE REVIEW

File: magma/backend/mlir/xmr_utils.py [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+    if isinstance(ref, TupleRef):
+        path = _ascend_to_leaf(ref.tuple, leaves)
+        return path + [f".{ref.index}"]
+    if isinstance(ref, ArrayRef):
+        path = _ascend_to_leaf(ref.array, leaves)
+        return path + [f"[{ref.index}]"]
+    raise TypeError(ref, value)
+
+
+def _advance(it):
```
Comment (leonardt) [inline magma/backend/mlir/xmr_utils.py:56]: Dead code?

File: magma/backend/mlir/xmr_utils.py [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+    return seperator.join(path)
+
+
+def _get_leaf_descendants(tree, node, include_self=False):
+    is_leaf = True
+    for succ in tree.successors(node):
+        is_leaf = False
+        yield from _get_leaf_descendants(tree, succ, include_self=True)
+    if is_leaf and include_self:
+        yield node
```
Comment (leonardt) [inline magma/backend/mlir/xmr_utils.py:32]: Could just use succ as the flag
```suggestion
    succ = None
    for succ in tree.successors(node):
        yield from _get_leaf_descendants(tree, succ, include_self=True)
    if succ is not None and include_self:
        yield node
```

File: magma/backend/mlir/hardware_module.py [hunk 3]
```diff
@@ -613,6 +620,39 @@ def make_register(T, data, init, result):
...
+        try:
+            self._ctx.parent.xmr_paths[xmr]
+        except KeyError:
+            pass
+        else:
+            return True
+        paths = get_xmr_paths(self._ctx, xmr)
+        assert len(paths) == len(module.operands)
+        self._ctx.parent.xmr_paths[xmr] = paths
+        return True
```
Comment (leonardt) [inline magma/backend/mlir/hardware_module.py:638]: Could we just move the code inside KeyError?
```suggestion
        try:
            self._ctx.parent.xmr_paths[xmr]
        except KeyError:
            paths = get_xmr_paths(self._ctx, xmr)
            assert len(paths) == len(module.operands)
            self._ctx.parent.xmr_paths[xmr] = paths
        return True
```

File: magma/backend/mlir/xmr_utils.py [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+        leaves = _get_leaf_descendants(tree, xmr.port)
+        leaves = sorted(leaves, key=lambda n: tree.nodes[n]["index"])
+        return [
+            (_path_to_string(_get_path(tree, leaf), "_"),)
+            for leaf in leaves
+        ]
+    # (2)
+    leaves = list(_get_leaf_descendants(tree, root, include_self=True))
+    leaf, *path = _ascend_to_leaf(xmr.port, leaves)
+    path = _path_to_string(_get_path(tree, leaf), "_") + "".join(path)
```
Comment (leonardt) [inline magma/backend/mlir/xmr_utils.py:83]: Seems like we could use a helper here for repeated pattern.
```
def _make_path(tree, value):
    return _path_to_string(_get_path(tree, value), "_")
```

**Review by leonardt — APPROVED** [review]
Minor comments

File: magma/backend/mlir/hardware_module.py (hunk 3 above)
Comment (rsetaluri) [reply magma/backend/mlir/hardware_module.py:638]: semantically, yes I think that would be fine. Just a style preference to do it outside, with an early return in the no-exception case.

**Pull request merged**